src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -1380,9 +1469,47 @@ func (s *commitLogSource) ReadIndex(
 		}
 	}
 
+	shouldReturnUnfulfilled, err := s.shouldReturnUnfulfilled(
+		encounteredCorruptData, ns, shardsTimeRanges, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if shouldReturnUnfulfilled {
+		indexResult.SetUnfulfilled(shardsTimeRanges)
+	}
 	return indexResult, nil
 }
 
+// If we encountered any corrupt data and there is a possibility of the
+// peers bootstrapper being able to correct it, we want to mark the entire range
+// as unfulfilled so the peers bootstrapper can attempt a repair, but keep
+// the data we read from the commit log as well in case the peers
+// bootstrapper is unable to satisfy the bootstrap because all peers are
+// down or if the commitlog contained data that the peers do not have.
+func (s commitLogSource) shouldReturnUnfulfilled(
+	encounteredCorruptData bool,
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	opts bootstrap.RunOptions,
+) (bool, error) {
+	if !s.opts.ReturnUnfulfilledForCorruptCommitlogFiles() {
+		return false, nil
+	}
+
+	if !encounteredCorruptData {
+		return false, nil
+	}
+
+	couldObtainDataFromPeers, err := s.couldObtainDataFromPeers(
```

Comment:
hm - i thought you were going to check if it's a single node for this case. why use the more extended peer condition?

Reply:
Seemed like it was slightly better to not hand off to the peers bootstrapper if it can't help anyways. Think I should remove it?